c2rust-analyze/src/dataflow/type_check.rs

```diff
@@ -199,21 +199,29 @@ impl<'tcx> TypeChecker<'tcx, '_> {
         }
     }
 
-    /// Unify corresponding `PointerId`s in `lty1` and `lty2`.
+    /// Unify corresponding [`PointerId`]s in `pl_lty` and `rv_lty`.
     ///
-    /// The two inputs must have compatible ([safely transmutable](are_transmutable)) underlying types.
+    /// The two inputs must have compatible ([safely transmutable](is_transmutable_to)) underlying types.
     /// For any position where the underlying type has a pointer,
-    /// this function unifies the `PointerId`s that `lty1` and `lty2` have at
-    /// that position.  For example, given `lty1 = *mut /*l1*/ *const /*l2*/ u8` and `lty2 = *mut
-    /// /*l3*/ *const /*l4*/ u8`, this function will unify `l1` with `l3` and `l2` with `l4`.
-    fn do_unify(&mut self, lty1: LTy<'tcx>, lty2: LTy<'tcx>) {
-        let ty1 = lty1.ty;
-        let ty2 = lty2.ty;
-        assert!(are_transmutable(
-            self.acx.tcx().erase_regions(ty1),
-            self.acx.tcx().erase_regions(ty2),
-        ), "types not transmutable (compatible), so PointerId unification cannot be done: {ty1:?} !~ {ty2:?}");
-        for (sub_lty1, sub_lty2) in lty1.iter().zip(lty2.iter()) {
+    /// this function unifies the [`PointerId`]s that `pl_lty` and `rv_lty` have at that position.
+    /// For example, given
+    ///
+    /// ```
+    /// # fn(
+    /// pl_lty: *mut /*l1*/ *const /*l2*/ u8,
+    /// rv_lty: *mut /*l3*/ *const /*l4*/ u8,
+    /// # ) {}
+    /// ```
+    ///
+    /// this function will unify `l1` with `l3` and `l2` with `l4`.
+    fn do_unify(&mut self, pl_lty: LTy<'tcx>, rv_lty: LTy<'tcx>) {
```

Comment:
i think we want to abstract this away from implying that unification is only from the perspective of a MIR assignment, so I don't think the `pl_lty` and `rv_lty` names will always be appropriate

Reply:
on second glance, it does look like it's only done in the context of some sort of assignment. so maybe it's fine. i'm not sure where we may want to do equivalence relations outside of assignment, maybe @spernsteiner can say

Reply:
Yeah, I looked at the call sites and they're always `pl_lty` and `rv_lty`, too, which is why I named it that.  I can also rename it to `from_lty` and `to_lty`, similar to the `is_transmutable_to` arg names.